### ISSUE TITLE
Extract multimedia information

### ISSUE THREAD

**Issue by bartaelterman** [Datafable/gbif-dataset-metrics#63]
## Description

We want to display **multimedia categories** (how many occurrences have multimedia) and a **sample of images** that are referred to in multimedia.txt. For multimedia categories (previously documented in #61 ), we need sums of occurrences per dataset while for the sample of images (previously documented in #60), we need urls to multimedia and the id of the corresponding occurrence.

## Required Output

In order to satisfy the two requirements, we need the following output:

```JS
{
    "datasetKey": {
        "movingimage": {
            "occurrenceID1": [url1, url2, url3],
            "occurrenceID2": [url1, url2]
        },
        "audio": {
            "occurrenceID1": [url1, url2],
            "occurrenceID2": [url1]
        },
        "stillimage": {
            "occurrenceID1": [url1, url2, url3, url4],
            "occurrenceID2": [url1, url2]
        },
        "no type": {
            "occurrenceID1": [url1],
            "occurrenceID2": [url1]
        },
        "media_not_provided": 37,
        "media_url_invalid": 12,
        "media_valid": 61
    }
}
```

Note that the sum of `media_not_provided`, `media_url_invalid` and `media_valid` should equal the number of occurrences for this `datasetKey`.

## Terms we need

```
id (occurrence.txt)
issues (occurrence.txt)
type (multimedia.txt)
references (multimedia.txt)
```

## Process

We'll loop over the occurrences, and expect their multimedia.

```JS
audio = {}
stillimage = {}
movingimage = {}
notype = {}

media_valid = 0
media_not_provided = 0
media_url_invalid = 0

// Loop over occurrence.txt
FOR EACH occurrence
    IF issues CONTAINS ( MULTIMEDIA_URI_INVALID )
        THEN media_url_invalid += 1
        // This issue should get priority over valid multimedia
        // This also prevents adding invalid url to the resources
    ELSE search for related multimedia
         IF found
             // no matter what the multimedia type, add a count to media_valid
             THEN media_valid += 1
             FOR EACH multimedia record
             // add every url to a multimedia resource
                 IF type is one of [audio, stillimage, movingimage]
                     THEN add id: array(references) to audio, stillimage or movingimage
                 ELSE
                     THEN add id: array(references) to notype
         ELSE no related media found
             THEN media_not_provided += 1
```

**Comment by peterdesmet** [comment]
@bartaelterman, do you know what is required to make this work or is this functional already?

**Comment by bartaelterman** [comment]
I'll have to examine this further. Looking at [the code](https://github.com/Datafable/gbif-dataset-metrics/blob/master/aggregation_module/src/aggregator.py#L38) it looks like the data extraction module provides the media urls but the aggregator skips it.

**Comment by peterdesmet** [comment]
@bartaelterman, looking in the CartoDB table, I see that `images_sample` is currently populated with `null`, the string `"null"` or an object with images. The string should be avoided, but the frontend code can handle it.

The format I currently get is:
```
{"501938848": "http://www.whaleshark.org:80/encounters/encounter.jsp?number=1692011171319", "501943513": "http://www.whaleshark.org/encounters/encounter.jsp?number=1511201011287", "501943534": "http://www.whaleshark.org/encounters/encounter.jsp?number=1811201083729", "501943492": "http://www.whaleshark.org/encounters/encounter.jsp?number=29620103586", "501943484": "http://www.whaleshark.org/encounters/encounter.jsp?number=296201041923", "501936762": "http://www.whaleshark.org:80/encounters/encounter.jsp?number=199201192935"}
```

Will this remain the same?

**Comment by peterdesmet** [comment]
Looking into the data I currently get back, we should only populate `images_sample` with `stillimage`. I currently have many false positives (e.g. [sounds](http://www.gbif.org/dataset/bc7c19ec-eefd-403b-931a-16a2ea1e71fc)). The fact that the URL is currently populated with the homepage URL (see #86), makes it harder to test.

**Comment by bartaelterman** [comment]
@peterdesmet what is the dataset key of that dataset?

About the `stillimage` I agree, but I think we discussed this before. For many datasets, the pictures are in `notype`. I think we agreed to use `stillimage`unless they are empty, and only in that case select images (check for the right extension maybe?) in `notype`.

**Comment by peterdesmet** [comment]
The dataset with sounds is: http://www.gbif.org/dataset/bc7c19ec-eefd-403b-931a-16a2ea1e71fc.

So currently it is checking for `stillimage` and then looking `notype` if that is empty? That's good to know for the documentation. Can you send me a sample of datasets with only data in `notype`? I can then assess how many positive hits we get and decided if it's worth implementing as such or only taking `stillimage`.

**Comment by bartaelterman** [comment]
Here's one: `50c9509d-22c7-4a22-a47d-8c48425ef4a7`

`stillimage` is not really empty, but there are no valid urls in it. The actual links to images are in `no_type` (with underscore).

**Comment by bartaelterman** [comment]
Should've checked the code first...

The aggregator checks whether the `stillimage` object is empty. If so, it uses the `no_type` object. However, the `stillimage` object is not empty, it just doesn't contain valid urls. So in this case, you probably won't find the valid urls in the output of the aggregator.

**Comment by peterdesmet** [comment]
So, do you have a dataset where `stillimage` is truly empty, and `no_type` was used? Also, in the extractor part, I assume `associatedMedia` is checked if no multimedia extention is available... and if there's content there, that will be added to `no_type`?

**Comment by peterdesmet** [comment]
So, I checked the 45 datasets (in the old data) with data in `images_sample`. This is what I discovered:

* Apparently the extractor looks in 3 fields for URLs, in this order: `references` (which it should never do), `identifier` (which is the good one) and `associatedMedia`.
* 25 datasets use the extension, of which 17 reference an image file in `identifier`. 14 of those have a `format` that indicates an image file (e.g. tiff, image/jpeg, etc.). @niconoe, are you getting nicer types back by GBIF (e.g. stillImage, etc.)? If we don't check against format, we get 25-14 = 9 false positives (sounds, web pages).
* 17 datasets use `associatedMedia`, but all except 2 don't reference image files, but pages.
* For 3 datasets I couldn't check because the occurrence IDs are no longer valid.

That leads me to the conclusion that we should only consider the group of `stillImages`, to avoid the 15 webpages in associatedMedia and 9 of no or other type in the extension. Too bad this will include the valid use of associatedMedia in this dataset, which I helped to publish: http://www.gbif.org/dataset/e8a25a42-f125-476c-8554-3ec21cd51a84
